src/test/kotlin/com/pubnub/api/subscribe/eventengine/effect/HandshakeReconnectEffectTest.kt

```diff
@@ -0,0 +1,136 @@
+import com.pubnub.api.subscribe.eventengine.event.Event
+import com.pubnub.api.subscribe.eventengine.event.SubscriptionCursor
+import io.mockk.every
+import io.mockk.mockk
+import io.mockk.verify
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Test
+import java.time.Duration
+import java.util.concurrent.Executors
+
+class HandshakeReconnectEffectTest {
+    private val channels = listOf("channel1")
+    private val channelGroups = listOf("channelGroup1")
+    private val reason = PubNubException("Unknown error")
+    private val attempts = 1
+    private val handshakeProvider = mockk<HandshakeProvider>()
+    private val eventDeliver = SubscribeManagedEffectFactoryTest.TestEventDeliver()
+    private val handshakeReconnectInvocation =
+        SubscribeEffectInvocation.HandshakeReconnect(channels, channelGroups, attempts, reason)
+    private val executorService = Executors.newSingleThreadScheduledExecutor()
+    private val policy = mockk<RetryPolicy>()
+    private val subscriptionCursor = SubscriptionCursor(1337L, "1337")
+
+    @Test
+    fun `should deliver HandshakeReconnectSuccess event when HandshakeReconnectEffect succeeded`() {
+        // given
+        every { handshakeProvider.handshake(any(), any()) } returns successfulRemoteAction(subscriptionCursor)
+        every { policy.nextDelay(any()) } returns Duration.ofMillis(10)
+        val handshakeReconnectEffect = HandshakeReconnectEffect(
+            handshakeProvider,
+            eventDeliver,
+            policy,
+            handshakeReconnectInvocation,
+            executorService
+        )
+
+        // when
+        handshakeReconnectEffect.runEffect()
+
+        // then
+        Thread.sleep(50)
+        assertEquals(
+            listOf(Event.HandshakeReconnectSuccess(channels, channelGroups, subscriptionCursor)),
+            eventDeliver.events
+        )
+    }
+
+    @Test
+    fun `should deliver HandshakeReconnectFailure event when HandshakeReconnectEffect failed`() {
+        // given
+        every { handshakeProvider.handshake(any(), any()) } returns failingRemoteAction(reason)
+        every { policy.nextDelay(any()) } returns Duration.ofMillis(10)
+        val handshakeReconnectEffect = HandshakeReconnectEffect(
+            handshakeProvider,
+            eventDeliver,
+            policy,
+            handshakeReconnectInvocation,
+            executorService
+        )
+
+        // when
+        handshakeReconnectEffect.runEffect()
+
+        // then
+        Thread.sleep(50)
+        assertEquals(listOf(Event.HandshakeReconnectFailure(reason)), eventDeliver.events)
+    }
+
+    @Test
+    fun `should deliver HandshakeReconnectGiveUp event when delay is null`() {
+        // given
+        every { policy.nextDelay(any()) } returns null
+        val handshakeReconnectEffect = HandshakeReconnectEffect(
+            handshakeProvider,
+            eventDeliver,
+            policy,
```

Comment:
You could use here something already written:

```suggestion
            NoRetriesPolicy,
```

Reply:
ok, I will use it.